### DISCUSSION TITLE
How to call custom scripts from other custom scripts the best way?

### DISCUSSION THREAD

**Discussion by zshorvat** [netbox-community/netbox#7395] [General]
Hi there,

We have started to develop a number of custom scripts to perform different actions (they work brilliantly, btw), but after all changes we should update the to-be-modified-devices' config context, for which we have a separate custom script.

I know that it is possible to call the custom scripts via API, but is there a way to call a script directly or at least import classes/functions from them?

Thanks a lot,
Zsolt

**Comment by DanSheps** [comment]
You will want to look for the enqueue_job (I think it is that) in the netbox code.  It is what handles sending the job to redis.

**Comment by jeremystretch** [comment]
While this may technically work, it calling one script from another certainly feels like an anti-pattern.

My recommendation would be to separate out whatever second phase logic you need into something that can be called by either this script or the other script directly.

**Comment by zshorvat** [comment]
Thanks for the superfast replies!

I was thinking about moving the class out completely, too, I think that makes most sense.

Is there a place where I should be placing my custom modules? I think I couldn't pick up files locally from the script folder.

Thanks again,
Zsolt

**Reply by DanSheps** [reply]
I have done this, class reuse  is definitely the better way to go

**Comment by hendrikbl** [comment]
Did you find out where to put custom modules? They will not be loaded if I simply put them in the scripts folder next to my custom scripts.

**Reply by DanSheps** [reply]
They should be, if they are not being loaded there is something wrong with the script.

I currently store my reuses in separate folders but I have also stored them in just a "helpers.py" file.  Anything in scripts should work.

**Reply by hendrikbl** [reply]
Okay, that's interesting. We're currently running netbox via [netbox-docker](https://github.com/netbox-community/netbox-docker), maybe that's causing any issues with how they run custom scripts.

I will ask over there and see what I find out.

**Reply by zshorvat** [reply]
Actually I had to move my module into the site-packages directory as I got errors when trying to pick it up from the script folder directly.

**Reply by kkthxbye-code** [reply]
By default netbox-docker mounts scripts like this: 
`- ./scripts:/etc/netbox/scripts:z,ro`

As that folder is not part of the path, it wont pick up files/folder in the same dir. You can add another mount in the override like so:
`./scripts:/opt/netbox/netbox/scripts:z,ro` 

and then change the following in netbox.env:
`SCRIPTS_ROOT=/opt/netbox/netbox/scripts`

Alternatively you can change the PATH in the script with something like this:
```python
import os, sys
sys.path.append(os.path.dirname(__file__))
```

But that solution is kinda icky.

There might be better solutions though.

**Reply by hendrikbl** [reply]
Neither of these solutions seems to work.
And to be honest I don't really get why making the scripts available through PATH should be necessary. From my (to be fair limited) python experience it should be enough to place a .py file next to the script that gets loaded.

Or are netbox scripts run in a different context where the surrounding files are not present?

**Reply by kkthxbye-code** [reply]
How are you importing the .py file?

If you have a util.py file in the script root, you need to import it like so if you changed the script path to what I suggested:
`from scripts import util`

>Or are netbox scripts run in a different context where the surrounding files are not present?

They are imported like so:

https://github.com/netbox-community/netbox/blob/develop/netbox/extras/scripts.py#L479-L492

When running in netbox-docker they are also executed on another container, not that it matters a lot in the case. When the scripts are loaded from a dictionary not in the path, the path is not automatically appended. So while the main script loads fine because of the logic i linked, when you do "import util" it just looks for util.py in the path, which does not contain /etc/netbox/scripts for example.

**Reply by hendrikbl** [reply]
Okay, it's starting to make sense now.
I imported my scripts with `import utils` which following your explanation doesn't work.
With your suggestions `from scripts import utils` and the new volume mount  `./scripts:/opt/netbox/netbox/scripts:z,ro` it works like a charm. Thanks :)

**Reply by DanSheps** [reply]
> By default netbox-docker mounts scripts like this: `- ./scripts:/etc/netbox/scripts:z,ro`
> 
> As that folder is not part of the path, it wont pick up files/folder in the same dir. You can add another mount in the override like so: `./scripts:/opt/netbox/netbox/scripts:z,ro`

This doesn't make sense

How does this look in the file system?  I suspect it is actually as a mount point.  If that is the case, is it possible that it is missing __init__.py in the scripts folder?

**Reply by kkthxbye-code** [reply]
Not sure what you mean, could you elaborate on what doesn't make sense. Docker volumes are not mount points.

In the container the scripts are in /etc/netbox/scripts and the SCRIPTS_ROOT setting is set to that folder. `__init__.py` doesn't make any differences and has been present in all my testing.

This is the path printed from a script run from /etc/netbox/scripts:

`['/opt/netbox/netbox', '/usr/lib/python39.zip', '/usr/lib/python3.9', '/usr/lib/python3.9/lib-dynload', '/opt/netbox/venv/lib/python3.9/site-packages']`

`/opt/netbox/netbox` is of course the netbox source, which is why `from scripts import util` works. To do relative imports from `/etc/netbox/scripts` one would have to add it to the path.

**Reply by DanSheps** [reply]
Sounds like the easiest way would be to add /etc/netbox/scripts to the path.

**Reply by zshorvat** [reply]
That was indeed the solution for me as well - adding the scripts folder to path.
